src/Core/SettingsEnums.h

```diff
@@ -42,9 +42,10 @@ enum class JoinAlgorithm
     HASH,
     PARTIAL_MERGE,
     PREFER_PARTIAL_MERGE,
+    DIRECT,
 };
 
-DECLARE_SETTING_ENUM(JoinAlgorithm)
+DECLARE_SETTING_MULTI_ENUM(JoinAlgorithm)
```

Comment:
Not sure if we should make this multi enum, it seems that validation becomes much harder. Maybe this DIRECT should be part of another join setting enum ?

Reply:
It's one of the things to consider. My motivation was the following: we have several join algorithms and will have more in the future. And for now, we can't choose from them automatically. But we can allow the user to specify the list of algorithms that ClickHouse can use and which cannot (for example, if the user knows that some of them wouldn't work on his data or disable some unstable algorithm). And we can try to choose from the available list; if storage supports kv, use `direct`; if not, choose another allowed by the user (it may be `hash` or `partial_merge`). If a user wants to force `direct` for a particular request, specify only it. And it saves us from introducing many new boolean flags for different algorithms.

Reply:
Agree it make sense.